File: cards.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
)

func main() {
	var hand Hand
	n_flush := 0
	n_straight := 0
	n_straight_flush := 0
	n_hands := 1_000_000
	n_pair, n_twopair, n_threes, n_fours, n_fullhouse := 0, 0, 0, 0, 0
	for i := 0; i < n_hands; i++ {
		hand = random_hand()
		if hand.IsStraight() {
			n_straight += 1
		}
		if hand.IsFlush() {
			n_flush += 1
		}
		if hand.IsStraightFlush() {
			n_straight_flush += 1
		}
		if hand.IsPair() {
			n_pair += 1
		}
		if hand.IsTwoPair() {
			n_twopair += 1
		}
		if hand.IsThreeOfAKind() {
			n_threes += 1
		}
		if hand.IsFourOfAKind() {
			n_fours += 1
		}
		if hand.IsFullHouse() {
			n_fullhouse += 1
		}
		if false {
			fmt.Println(hand)
		}
	}
	fmt.Printf("%v hands\n", n_hands)
	fmt.Printf("%v pair => %.2f%%\n", n_pair, percent(n_pair, n_hands))
	fmt.Printf("%v two_pair => %.2f%%\n", n_twopair, percent(n_twopair, n_hands))
	fmt.Printf("%v three_of => %.2f%%\n", n_threes, percent(n_threes, n_hands))
	fmt.Printf("%v straight => %.2f%%\n", n_straight, percent(n_straight, n_hands))
	fmt.Printf("%v flush => %.2f%%\n", n_flush, percent(n_flush, n_hands))
	fmt.Printf("%v house => %.2f%%\n", n_fullhouse, percent(n_fullhouse, n_hands))
	fmt.Printf("%v four_of => %.2f%%\n", n_fours, percent(n_fours, n_hands))
	fmt.Printf("%v straight flush => %.2f%%\n", n_straight_flush, percent(n_straight_flush, n_hands))
}

func percent(num, denom int) float64 {
	return float64(num) / float64(denom) * 100
}

type Card struct {
	suit byte
	num  int
}

func (c Card) String() string {
	return fmt.Sprintf("%s%d", string(c.suit), c.num)
}

type Hand struct {
	hand    [5]Card
	cardMap map[int]int
}

func (h Hand) String() string {
	return fmt.Sprintf("<%s %s %s %s %s>",
		h.hand[0],
		h.hand[1],
		h.hand[2],
		h.hand[3],
		h.hand[4],
	)
}

func (h *Hand) MakeCardMap() {
	nums := make(map[int]int)
	for _, card := range h.hand {
		nums[card.num] += 1
	}
	h.cardMap = nums
}

func (h Hand) IsStraight() bool {
	var nums = make([]int, 5)
	for i, card := range h.hand {
		nums[i] = card.num
	}
	sort.Ints(nums)
	prev := nums[0]
	for _, num := range nums[1:] {
		if num != prev+1 {
			return false
		}
		prev = num
	}
	return true
}

func (h Hand) IsFlush() bool {
	first := h.hand[0].suit
	for _, card := range h.hand[1:] {
		if card.suit != first {
			return false
		}
	}
	return true
}

func (h Hand) IsStraightFlush() bool {
	return h.IsStraight() && h.IsFlush()
}

func (h Hand) IsPair() bool {

	found_pair := false
	for _, n_num := range h.cardMap {
		if n_num == 2 && !found_pair {
			found_pair = true
		} else if n_num == 2 {
			return false
		}
	}
	return found_pair
}

func (h Hand) IsTwoPair() bool {
	found_pair := false
	for _, n_num := range h.cardMap {
		if n_num == 2 && !found_pair {
			found_pair = true
		} else if n_num == 2 {
			return true
		}
	}
	return false
}

func (h Hand) IsThreeOfAKind() bool {
	for _, n_num := range h.cardMap {
		if n_num == 3 {
			return true
		}
	}
	return false
}

func (h Hand) IsFourOfAKind() bool {
	for _, n_num := range h.cardMap {
		if n_num == 4 {
			return true
		}
	}
	return false
}

func (h Hand) IsFullHouse() bool {
	return h.IsPair() && h.IsThreeOfAKind()
}

func random_hand() Hand {
	suits := []byte{'C', 'H', 'S', 'D'}

	var cards [5]Card
	// fmt.Println(num)

	n := 0
	for {
		rand_suit := suits[rand.Intn(4)]
		rand_card := rand.Intn(13)
		temp := Card{rand_suit, rand_card}
		for _, card := range cards {
			if card == temp {
				continue
			}
		}
		cards[n] = temp
		n += 1
		if n == 5 {
			break
		}
	}
	h := Hand{cards, make(map[int]int)}
	h.MakeCardMap()
	return h
}

File: cards.py
#!/usr/bin/env python3
import numpy as np
import random
from collections import Counter
from itertools import islice
import sys

from dataclasses import dataclass
from typing import List, Dict


random.seed(None)
ALL_CARDS = [(suit, number) for number in range(1, 14) for suit in "CHSD"]


@dataclass
class Hand:
    def __init__(self, cards):
        self.hand = cards
        self.suits = Counter(s for s, n in cards)
        self.nums = Counter(int(n) for s, n in cards)
        self.number_counts = list(self.nums.values())
        self.n_unique_nums = len(self.nums.keys())

    def add(self, other):
        assert len(self.hand) < 5, "Can't have more than 5 cards"
        if isinstance(other, list):
            assert len(self.hand) + len(other) <= 5, "Can't have more than 5 cards"
            self.hand.extend(other)
        else:
            assert len(self.hand) <= 4, "Can't have more than 5 cards"
            self.hand.extend([other])
        self.suits = Counter(s for s, n in self.hand)
        self.nums = Counter(int(n) for s, n in self.hand)
        self.number_counts = list(self.nums.values())
        self.n_unique_nums = len(self.nums.keys())

    @staticmethod
    def random():
        return Hand(random.choices(ALL_CARDS, k=5))

    def __str__(self):
        return "Hand(" + " ".join(f"{suit}{number}" for suit, number in self.hand) + ")"

    def __repr__(self):
        return self.__str__()


    def is_full_house(self):
        return self.n_unique_nums == 2 and 3 in self.number_counts

    def is_flush(self):
        return len(self.suits.keys()) == 1

    def is_straight_flush(self):
        return self.is_flush() and self.is_straight()

    def is_straight(self):
        nums = sorted(self.nums)
        if len(nums) < 5:
            return False
        return nums[0] == nums[1] - 1 == nums[2] - 2 == nums[3] - 3 == nums[4] - 4

    def is_pair(self):
        return self.n_unique_nums == 4 and 2 in self.number_counts

    def is_twopair(self):
        nums = list(self.number_counts)
        if 2 in nums:
            nums.remove(2)
        return 2 in nums

    def is_three_of_a_kind(self):
        return self.n_unique_nums == 3 and 3 in self.number_counts

    def is_four_of_a_kind(self):
        return self.n_unique_nums == 2 and 4 in self.number_counts


def random_hand_generator(n=None):
    generated = 0
    while True:
        yield Hand.random()
        generated += 1
        if n and generated == n:
            break


def prob_of_future_hands(current):
    remaining = remaining_cards(current)
    print(len(remaining))
    probs = {
        "pair": prob_getting_pair(current),
        "two_pair": prob_getting_two_pair(current),
        "three_of_a_kind": prob_getting_three_of_a_kind(current),
        "four_of_a_kind": prob_getting_four_of_a_kind(current),
        "full_house": prob_getting_full_house(current),
        "flush": prob_getting_flush(current),
        "straight": prob_getting_straight(current),
        "straight_flush": prob_getting_straight_flush(current),
    }


def remaining_cards(current):
    all_cards = ALL_CARDS.copy()
    # print(len(all_cards))
    for card in current.hand:
        all_cards.remove(card)
    return all_cards


def prob_getting_pair(current):
    # if current.is_pair():
    #     return 1
    # else:
    #     if len(current.hand) == 5:
    #         return 0
    #     remaining = remaining_cards(current)
    #     return prob_getting_pair()
    pass


def prob_getting_two_pair(current):
    pass


def prob_getting_three_of_a_kind(current):
    pass


def prob_getting_four_of_a_kind(current):
    pass


def prob_getting_full_house(current):
    pass


def prob_getting_flush(current):
    suits = [s for s, _ in current.hand]
    if not all([s == suits[0] for s in suits]):
        return 0
    if len(suits) == 5:
        return 1
    remaining = remaining_cards(current)
    remaining_suit = [r for r in remaining if r[0] == suits[0]]
    diff = 5 - len(current.hand)
    return (len(remaining_suit) - diff - 1) / len(remaining)


def prob_getting_straight(current):
    pass


def prob_getting_straight_flush(current):
    suits = [s for s, _ in current]
    if not all([s == suits[0] for s in suits]):
        return 0
    pass


# gen = random_hand_generator()
# hand = next(gen)
# hand = Hand([("C", 12), ("H", 12)])
# print(hand)
# print(prob_of_future_hands(hand))

# print(prob_getting_flush(hand))
# print(prob_getting_flush(Hand([("H", 1)])))

how_many_hands = 1_000_000
n_pairs = 0
n_twos = 0
n_threes = 0
n_fours = 0
n_full = 0
n_straight = 0
n_straightflush = 0
n_flush = 0
percents = list(np.linspace(0, how_many_hands, 11))

rng = np.random.default_rng()
choices = rng.choice(ALL_CARDS, (5, 5))
hands = [Hand(choices.tolist()) for choices in rng.choice(ALL_CARDS, (how_many_hands, 5))]

n_pairs = sum(1 for hand in hands if hand.is_pair())
n_twos = sum(1 for hand in hands if hand.is_twopair())
n_threes = sum(1 for hand in hands if hand.is_three_of_a_kind())
n_fours = sum(1 for hand in hands if hand.is_four_of_a_kind())
n_flush = sum(1 for hand in hands if hand.is_flush())
n_straight = sum(1 for hand in hands if hand.is_straight())
n_straightflush = sum(1 for hand in hands if hand.is_straight_flush())
n_full = sum(1 for hand in hands if hand.is_full_house())

for hand in hands:
    if hand.is_pair():
        n_pairs += 1
    if hand.is_twopair():
        n_twos += 1
    if hand.is_three_of_a_kind():
        n_threes += 1
    if hand.is_four_of_a_kind():
        n_fours += 1
    if hand.is_flush():
        n_flush += 1
    if hand.is_straight():
        n_straight += 1
    if hand.is_straight_flush():
        n_straightflush += 1
    if hand.is_full_house():
        n_full += 1

print("\r", end="")
print(f"Pair:           {n_pairs/how_many_hands * 100:.2f}%")
print(f"Two Pair:       {n_twos/how_many_hands * 100:.2f}%")
print(f"Threes:         {n_threes/how_many_hands * 100:.2f}%")
print(f"Full House:     {n_full/how_many_hands * 100:.2f}%")
print(f"Four:           {n_fours/how_many_hands * 100:.2f}%")
print(f"Flush:          {n_flush/how_many_hands * 100:.2f}%")
print(f"Straight:       {n_straight/how_many_hands * 100:.2f}%")

